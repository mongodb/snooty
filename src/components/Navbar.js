import React, { useCallback, useEffect, useMemo, useState } from 'react';
import useMedia from '../hooks/use-media';
import { withPrefix } from 'gatsby';
import styled from '@emotion/styled';
import { isBrowser } from '../utils/is-browser';
import { getSearchbarResultsFromJSON } from '../utils/get-searchbar-results-from-json';
import { searchParamsToURL } from '../utils/search-params-to-url';
import { URL_SLUGS } from '../constants';
import Searchbar from './Searchbar';

const getActiveSection = (slug, urlItems) => {
  const urlMapping = Object.entries(urlItems).find(([, value]) => value.includes(slug));
  if (urlMapping) {
    return urlMapping[0];
  }

  if (isBrowser) {
    switch (window.location.pathname) {
      case 'tools':
        return 'tools';
      case 'cloud':
        return 'cloud';
      default:
        return null;
    }
  }

  return null;
};

const NavbarContainer = styled('div')`
  ${({ isExpanded, shouldOpaqueWhenExpanded }) => isExpanded && shouldOpaqueWhenExpanded && 'opacity: 0.2;'};
`;

const Navbar = () => {
  const [activeLink, setActiveLink] = useState('');
  // We want to expand the searchbar on default when it won't collide with any other nav elements
  // Specifically, the upper limit works around the Get MongoDB link
  const isSearchbarDefaultExpanded = useMedia(
    'only screen and (min-width: 670px) and (max-width: 1200px), (min-width: 1300px)'
  );
  const isActiveLink = useCallback(link => link.toLowerCase() === activeLink, [activeLink]);
  const [isSearchbarExpanded, setIsSearchbarExpanded] = useState(isSearchbarDefaultExpanded);
  const modifyActiveLink = useMemo(
    () =>
      `{"links": [
        {"url": "https://docs.mongodb.com/manual/","text": "Server", "active": ${isActiveLink('Server')}},
        {"url": "https://docs.mongodb.com/drivers/","text": "Drivers", "active": ${isActiveLink('Drivers')}},
        {"url": "https://docs.mongodb.com/cloud/","text": "Cloud", "active": ${isActiveLink('Cloud')}},
        {"url": "https://docs.mongodb.com/tools/","text": "Tools", "active": ${isActiveLink('Tools')}},
        {"url": "https://docs.mongodb.com/guides/","text": "Guides", "active": ${isActiveLink('Guides')}}
      ]}`,
    [isActiveLink]
  );
  const [navprops, setNavprops] = useState(modifyActiveLink);

  const onSearchbarExpand = useCallback(
    isExpanded => {
      // On certain screens the searchbar is never collapsed
      if (!isSearchbarDefaultExpanded) {
        setIsSearchbarExpanded(isExpanded);
      }
    },
    [isSearchbarDefaultExpanded]
  );

  useEffect(() => {
    // Add script to give navbar functionality and css
    const script = document.createElement('script');
    script.src = withPrefix('docs-tools/navbar.min.js');
    script.async = true;

    document.body.appendChild(script);

    setActiveLink(getActiveSection(process.env.GATSBY_SITE, URL_SLUGS));
  }, []);

  useEffect(() => {
    setNavprops(modifyActiveLink);
  }, [activeLink, modifyActiveLink]);

  useEffect(() => {
    setIsSearchbarExpanded(isSearchbarDefaultExpanded);
  }, [isSearchbarDefaultExpanded]);

  return (
    <>
      <NavbarContainer
        isExpanded={isSearchbarExpanded}
        shouldOpaqueWhenExpanded={!isSearchbarDefaultExpanded}
        tabIndex="0"
        id="navbar"
        className="navbar"
        data-navprops={navprops}
      />
      <Searchbar
        getResultsFromJSON={getSearchbarResultsFromJSON}
        isExpanded={isSearchbarExpanded}
        setIsExpanded={onSearchbarExpand}
        searchParamsToURL={searchParamsToURL}
        // Autofocus the searchbar when the user expands only so the user can start typing
        shouldAutofocus={!isSearchbarDefaultExpanded}
      />
    </>
  );
};

export default Navbar;
