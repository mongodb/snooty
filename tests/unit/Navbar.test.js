import React from 'react';
import { shallow } from 'enzyme';
import Navbar from '../../src/components/Navbar';

it('renders correctly without browser', () => {
  const tree = shallow(<Navbar />);
  expect(tree).toMatchSnapshot();
});