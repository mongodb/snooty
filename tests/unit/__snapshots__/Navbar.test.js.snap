// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without browser 1`] = `
<Fragment>
  <div
    className="navbar"
    data-navprops="{\\"links\\": [
                    {\\"url\\": \\"https://docs.mongodb.com/manual/\\",\\"text\\": \\"Server\\", \\"active\\": false},
                    {\\"url\\": \\"https://docs.mongodb.com/drivers/\\",\\"text\\": \\"Drivers\\", \\"active\\": false},
                    {\\"url\\": \\"https://docs.mongodb.com/cloud/\\",\\"text\\": \\"Cloud\\", \\"active\\": false},
                    {\\"url\\": \\"https://docs.mongodb.com/tools/\\",\\"text\\": \\"Tools\\", \\"active\\": false},
                    {\\"url\\": \\"https://docs.mongodb.com/guides/\\",\\"text\\": \\"Guides\\", \\"active\\": false}
    ]}"
    id="navbar"
    tabIndex="0"
  />
  <Searchbar />
</Fragment>
`;
